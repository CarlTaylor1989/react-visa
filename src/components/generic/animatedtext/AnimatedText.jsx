import React from 'react';
import PropTypes from 'prop-types';
import { Translate } from 'react-redux-i18n';

import './AnimatedText.scss';

const AnimatedText = ({ texts, playAnimation, handleAnimationEnded }) => (
  <div className={`animatedText${playAnimation ? ' play' : ''}`}>
    {
      Object.keys(texts).map((text, i) => (
        <div key={texts[text].id} className="text-container">
          <div
            className="text"
            onAnimationEnd={i === Object.keys(texts).length - 1 ? handleAnimationEnded : null}
          >
            <Translate dangerousHTML value={texts[text].text} />
          </div>
        </div>
      ))
    }
  </div>
);

AnimatedText.propTypes = {
  texts: PropTypes.array.isRequired,
  playAnimation: PropTypes.bool.isRequired,
  handleAnimationEnded: PropTypes.func.isRequired,
};

export default AnimatedText;
