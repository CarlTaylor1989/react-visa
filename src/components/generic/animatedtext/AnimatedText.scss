@import '../../../styles/mixins';
@import '../../../styles/variables';

.animatedText {
  height: 100px;
  position: relative;

  .text-container {
    align-items: center;
    display: flex;
    height: 100%;
    left: 0;
    position: absolute;
    text-align: center;
    top: 0;
    width: 100%;
    z-index: 1;

    .text {
      @include opacity(0);
      color: $colour-yellow;
      font-size: 60px;
      font-weight: bold;
      margin-bottom: 45px;
      text-transform: uppercase;
      width: 100%;
    }
  }

  &.play {
    .text-container {
      &:first-child {
        .text {
          @include animation(intro 4s 3s);
        }
      }

      &:nth-child(2) {
        .text {
          @include animation(intro 2s 7s);
        }
      }

      &:nth-child(3) {
        .text {
          @include animation(intro 3s 9s);
        }
      }

      &:last-child {
        .text {
          @include animation(introEnd 1s ease-out 12s normal forwards);
        }
      }
    }
  }

  @keyframes intro {
    0% {
      @include animation-timing-function(cubic-bezier(.2, .655, .4, .885));
      @include opacity(0);
      @include transform(perspective(2000px) translate3d(0, -50%, -1300px));
    }

    50% {
      @include animation-timing-function(cubic-bezier(.4, .015, .595, .99));
      @include opacity(1);
      @include transform(perspective(2000px) translate3d(0, -50%, 0));
    }

    100% {
      @include animation-timing-function(cubic-bezier(.735, .045, .95, .5));
      @include opacity(0);
      @include transform(perspective(2000px) translate3d(0, -50%, 1300px));
    }
  }

  @keyframes introEnd {
    0% {
      @include opacity(0);
      @include transform(perspective(2000px) translate3d(0, -50%, -100px));
    }

    100% {
      @include opacity(1);
      @include transform(perspective(2000px) translate3d(0, -50%, 0));
    }
  }
}
