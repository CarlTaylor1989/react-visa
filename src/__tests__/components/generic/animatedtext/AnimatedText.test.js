/* global toJson */
import React from 'react'; // eslint-disable-line no-unused-vars
import AnimatedText from '../../../../components/generic/animatedtext/AnimatedText';

let mockProps;
beforeEach(() => {
  jest.resetAllMocks();
  mockProps = {
    texts: [
      {
        id: 'one',
        text: 'MESSAGE 1 MOTIVATIONAL MESSAGE'
      },
      {
        id: 'two',
        text: 'MESSAGE 2 MOTIVATIONAL MESSAGE'
      },
      {
        id: 'three',
        text: 'MESSAGE 3 MOTIVATIONAL MESSAGE'
      },
      {
        id: 'four',
        text: 'MESSAGE 4 MOTIVATIONAL MESSAGE'
      }

    ],
    playAnimation: false,
    handleAnimationEnded: jest.fn()
  };
});

test('renders the AnimatedText component', () => {
  const wrapper = shallow(<AnimatedText {...mockProps} />);
  expect(toJson(wrapper, { noKey: false })).toMatchSnapshot();
});

test('doesnt have "play" class by default', () => {
  const wrapper = shallow(<AnimatedText {...mockProps} />);
  expect(wrapper.hasClass('play')).toEqual(false);
});


test('has "play" class if "playAnimation" is true', () => {
  const wrapper = shallow(<AnimatedText {...mockProps} playAnimation />);
  expect(wrapper.hasClass('play')).toEqual(true);
});

test('handleAnimationEnded is not called if animationEnd has not been triggered', () => {
  shallow(<AnimatedText {...mockProps} />);
  expect(mockProps.handleAnimationEnded).not.toHaveBeenCalled();
});

test('calls handleAnimationEnded when animation has finished', () => {
  const wrapper = shallow(<AnimatedText {...mockProps} />);
  wrapper.find('.text').last().simulate('animationEnd');

  expect(mockProps.handleAnimationEnded).toHaveBeenCalled();
  expect(mockProps.handleAnimationEnded).toHaveBeenCalledTimes(1);
});
