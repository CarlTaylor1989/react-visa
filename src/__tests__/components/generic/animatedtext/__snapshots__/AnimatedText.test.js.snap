// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the AnimatedText component 1`] = `
<div
  className="animatedText"
>
  <div
    className="text-container"
    key="one"
  >
    <div
      className="text"
      onAnimationEnd={null}
    >
      <Translate
        dangerousHTML={true}
        tag="span"
        value="MESSAGE 1 MOTIVATIONAL MESSAGE"
      />
    </div>
  </div>
  <div
    className="text-container"
    key="two"
  >
    <div
      className="text"
      onAnimationEnd={null}
    >
      <Translate
        dangerousHTML={true}
        tag="span"
        value="MESSAGE 2 MOTIVATIONAL MESSAGE"
      />
    </div>
  </div>
  <div
    className="text-container"
    key="three"
  >
    <div
      className="text"
      onAnimationEnd={null}
    >
      <Translate
        dangerousHTML={true}
        tag="span"
        value="MESSAGE 3 MOTIVATIONAL MESSAGE"
      />
    </div>
  </div>
  <div
    className="text-container"
    key="four"
  >
    <div
      className="text"
      onAnimationEnd={[MockFunction]}
    >
      <Translate
        dangerousHTML={true}
        tag="span"
        value="MESSAGE 4 MOTIVATIONAL MESSAGE"
      />
    </div>
  </div>
</div>
`;
